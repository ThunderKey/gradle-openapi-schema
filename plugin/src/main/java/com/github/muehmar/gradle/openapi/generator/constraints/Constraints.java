package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class Constraints {
  private final Min min;
  private final Max max;
  private final DecimalMin decimalMin;
  private final DecimalMax decimalMax;
  private final Size size;
  private final Pattern pattern;
  private final Email email;

  private Constraints(
      Min min,
      Max max,
      DecimalMin decimalMin,
      DecimalMax decimalMax,
      Size size,
      Pattern pattern,
      Email email) {
    this.min = min;
    this.max = max;
    this.decimalMin = decimalMin;
    this.decimalMax = decimalMax;
    this.size = size;
    this.pattern = pattern;
    this.email = email;
  }

  public static Constraints empty() {
    return new Constraints(null, null, null, null, null, null, null);
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public Constraints withMin(Min min) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withMax(Max max) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withSize(Size size) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public void onMin(Consumer<Min> onMin) {
    Optional.ofNullable(min).ifPresent(onMin);
  }

  public void onMax(Consumer<Max> onMax) {
    Optional.ofNullable(max).ifPresent(onMax);
  }

  public void onDecimalMin(Consumer<DecimalMin> onDecimalMin) {
    Optional.ofNullable(decimalMin).ifPresent(onDecimalMin);
  }

  public void onDecimalMax(Consumer<DecimalMax> onDecimalMax) {
    Optional.ofNullable(decimalMax).ifPresent(onDecimalMax);
  }

  public void onSize(Consumer<Size> onSize) {
    Optional.ofNullable(size).ifPresent(onSize);
  }

  public void onPattern(Consumer<Pattern> onPattern) {
    Optional.ofNullable(pattern).ifPresent(onPattern);
  }

  public void onEmail(Consumer<Email> onEmail) {
    Optional.ofNullable(email).ifPresent(onEmail);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Constraints that = (Constraints) o;
    return Objects.equals(min, that.min)
        && Objects.equals(max, that.max)
        && Objects.equals(size, that.size)
        && Objects.equals(pattern, that.pattern)
        && Objects.equals(email, that.email);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max, size, pattern, email);
  }

  @Override
  public String toString() {
    return "Constraints{"
        + "min="
        + min
        + ", max="
        + max
        + ", size="
        + size
        + ", pattern="
        + pattern
        + ", email="
        + email
        + '}';
  }
}
