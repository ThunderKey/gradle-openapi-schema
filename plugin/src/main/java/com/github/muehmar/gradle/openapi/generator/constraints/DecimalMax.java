package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

public class DecimalMax {
  private final String value;
  private final boolean inclusiveMax;

  public DecimalMax(String value, boolean inclusiveMax) {
    this.value = value;
    this.inclusiveMax = inclusiveMax;
  }

  public static DecimalMax inclusive(String value) {
    return new DecimalMax(value, true);
  }

  public DecimalMax exclusiveMax() {
    return new DecimalMax(value, false);
  }

  public String getValue() {
    return value;
  }

  public boolean isInclusiveMax() {
    return inclusiveMax;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    DecimalMax that = (DecimalMax) o;
    return inclusiveMax == that.inclusiveMax && Objects.equals(value, that.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value, inclusiveMax);
  }

  @Override
  public String toString() {
    return "DecimalMax{" + "value='" + value + '\'' + ", inclusiveMax=" + inclusiveMax + '}';
  }
}
