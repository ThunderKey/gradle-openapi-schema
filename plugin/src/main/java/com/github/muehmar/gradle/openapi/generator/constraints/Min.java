package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

/** Minimum constraint of a property */
public class Min {
  private final long value;

  public Min(long value) {
    this.value = value;
  }

  public long getValue() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Min min1 = (Min) o;
    return value == min1.value;
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return "Min{" + "value=" + value + '}';
  }
}
