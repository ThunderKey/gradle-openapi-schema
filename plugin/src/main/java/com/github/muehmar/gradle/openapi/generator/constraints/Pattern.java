package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

public class Pattern {
  private final String value;

  public Pattern(String value) {
    this.value = value;
  }

  public String getPattern() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Pattern pattern = (Pattern) o;
    return Objects.equals(value, pattern.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return "Pattern{" + "value='" + value + '\'' + '}';
  }
}
