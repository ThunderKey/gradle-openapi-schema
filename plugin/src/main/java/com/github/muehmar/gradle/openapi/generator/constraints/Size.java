package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;

/** Size constraint of a property. Either {@code min} or {@code max} will be present (or both). */
public class Size {
  private final Integer min;
  private final Integer max;

  private Size(Integer min, Integer max) {
    this.min = min;
    this.max = max;
  }

  public static Size of(int min, int max) {
    return new Size(min, max);
  }

  public static Size ofMin(int min) {
    return new Size(min, null);
  }

  public static Size ofMax(int max) {
    return new Size(null, max);
  }

  public Optional<Integer> getMin() {
    return Optional.ofNullable(min);
  }

  public Optional<Integer> getMax() {
    return Optional.ofNullable(max);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Size size = (Size) o;
    return Objects.equals(min, size.min) && Objects.equals(max, size.max);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max);
  }

  @Override
  public String toString() {
    return "Size{" + "min=" + min + ", max=" + max + '}';
  }
}
