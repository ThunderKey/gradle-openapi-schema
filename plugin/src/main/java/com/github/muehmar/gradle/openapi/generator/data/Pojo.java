package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.Resolver;

import java.util.Objects;
import java.util.Optional;

public class Pojo {
  private final String key;
  private final String description;
  private final String suffix;
  private final PList<PojoMember> members;
  private final boolean isArray;

  public Pojo(
      String key, String description, String suffix, PList<PojoMember> members, boolean isArray) {
    this.key = key;
    this.description = Optional.ofNullable(description).orElse("");
    this.suffix = suffix;
    this.members = members;
    this.isArray = isArray;
  }

  public String getKey() {
    return key;
  }

  public String className(Resolver resolver) {
    return resolver.className(key + suffix);
  }

  public String getDescription() {
    return description;
  }

  public PList<PojoMember> getMembers() {
    return members;
  }

  public boolean isArray() {
    return isArray;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Pojo pojo = (Pojo) o;
    return isArray == pojo.isArray
        && Objects.equals(key, pojo.key)
        && Objects.equals(description, pojo.description)
        && Objects.equals(suffix, pojo.suffix)
        && Objects.equals(members, pojo.members);
  }

  @Override
  public int hashCode() {
    return Objects.hash(key, description, suffix, members, isArray);
  }

  @Override
  public String toString() {
    return "Pojo{"
        + "key='"
        + key
        + '\''
        + ", description='"
        + description
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", members="
        + members
        + ", isArray="
        + isArray
        + '}';
  }
}
