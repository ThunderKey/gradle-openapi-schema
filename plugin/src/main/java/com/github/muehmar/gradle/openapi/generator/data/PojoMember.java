package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.Resolver;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class PojoMember {
  private final Name name;
  private final String description;
  private final Type type;
  private final boolean nullable;

  public PojoMember(Name name, String description, Type type, boolean nullable) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.type = type;
    this.nullable = nullable;
  }

  public String getDescription() {
    return description;
  }

  public Name getTypeName(Resolver resolver) {
    return type.isEnum() ? resolver.enumName(name) : type.getFullName();
  }

  public Type getType() {
    return type;
  }

  public Name getName() {
    return name;
  }

  public boolean isNullable() {
    return nullable;
  }

  public boolean isRequired() {
    return !isNullable();
  }

  public Name getterName(Resolver resolver) {
    return resolver.getterName(name, type);
  }
  /**
   * The provided {@code code} is executed in case this type is an enum with the list of members in
   * the enum as arguments.
   */
  public void onEnum(Consumer<PList<String>> code) {
    type.onEnum(code);
  }

  public Name setterName(Resolver resolver) {
    return resolver.setterName(name);
  }

  public Name witherName(Resolver resolver) {
    return resolver.witherName(name);
  }

  public Name memberName(Resolver resolver) {
    return resolver.memberName(name);
  }

  public PList<String> getImports() {
    return type.getImports();
  }

  /**
   * Replaces a type of a member with another description and type.
   *
   * @param memberType If memberType is equally to the full name of the current type, the given
   *     description and type will be replaced in this member.
   * @param newDescription Description of the member which should be used.
   * @param newType Type of the member which should be used
   */
  public PojoMember replaceMemberType(Name memberType, String newDescription, Type newType) {
    if (type.getFullName().equals(memberType)) {
      return new PojoMember(name, newDescription, newType, nullable);
    } else {
      return this;
    }
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoMember that = (PojoMember) o;
    return nullable == that.nullable
        && Objects.equals(name, that.name)
        && Objects.equals(description, that.description)
        && Objects.equals(type, that.type);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, description, type, nullable);
  }

  @Override
  public String toString() {
    return "PojoMember{"
        + "key='"
        + name
        + '\''
        + ", description='"
        + description
        + '\''
        + ", type="
        + type
        + ", nullable="
        + nullable
        + '}';
  }
}
