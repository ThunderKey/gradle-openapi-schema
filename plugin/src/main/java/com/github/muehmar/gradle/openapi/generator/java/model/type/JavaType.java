package com.github.muehmar.gradle.openapi.generator.java.model.type;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.ParameterizedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.Type;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;

public interface JavaType {
  QualifiedClassName getQualifiedClassName();

  Nullability getNullability();

  JavaType withNullability(Nullability nullability);

  /**
   * Returns the qualified classnames used for this type, including the classes of possible type
   * parameters.s
   */
  PList<QualifiedClassName> getAllQualifiedClassNames();

  ParameterizedClassName getParameterizedClassName();

  /**
   * Returns true in case this class is a java array (not to be confused with the openapi
   * array-type).
   */
  boolean isJavaArray();

  Constraints getConstraints();

  <T> T fold(
      Function<JavaArrayType, T> onArrayType,
      Function<JavaBooleanType, T> onBooleanType,
      Function<JavaEnumType, T> onEnumType,
      Function<JavaMapType, T> onMapType,
      Function<JavaAnyType, T> onAnyType,
      Function<JavaNumericType, T> onNumericType,
      Function<JavaIntegerType, T> onIntegerType,
      Function<JavaObjectType, T> onObjectType,
      Function<JavaStringType, T> onStringType);

  default boolean isArrayType() {
    return fold(
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance,
        JavaArrayType.class::isInstance);
  }

  default boolean isMapType() {
    return fold(
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance,
        JavaMapType.class::isInstance);
  }

  default boolean isAnyType() {
    return fold(
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance,
        JavaAnyType.class::isInstance);
  }

  default boolean isIntegerType() {
    return fold(
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance,
        JavaIntegerType.class::isInstance);
  }

  default boolean isNumericType() {
    return fold(
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance,
        JavaNumericType.class::isInstance);
  }

  default boolean isObjectType() {
    return onObjectType().isPresent();
  }

  default Optional<JavaObjectType> onObjectType() {
    return fold(
        javaArrayType -> Optional.empty(),
        javaBooleanType -> Optional.empty(),
        javaEnumType -> Optional.empty(),
        javaMapType -> Optional.empty(),
        javaAnyType -> Optional.empty(),
        javaNumericType -> Optional.empty(),
        javaIntegerType -> Optional.empty(),
        Optional::of,
        javaStringType -> Optional.empty());
  }

  default PList<QualifiedClassName> getImports() {
    return getAllQualifiedClassNames()
        .filter(qualifiedClassName -> qualifiedClassName.getPackageName().isPresent())
        .filter(qualifiedClassName -> not(qualifiedClassName.isJavaLangPackage()));
  }

  default PList<String> getImportsAsString() {
    return getImports().map(QualifiedClassName::asString);
  }

  static JavaType wrap(Type type, TypeMappings typeMappings) {
    return type.fold(
        numericType -> JavaNumericType.wrap(numericType, typeMappings),
        numericType -> JavaIntegerType.wrap(numericType, typeMappings),
        stringType -> JavaStringType.wrap(stringType, typeMappings),
        arrayType -> JavaArrayType.wrap(arrayType, typeMappings),
        booleanType -> JavaBooleanType.wrap(booleanType, typeMappings),
        JavaObjectType::wrap,
        enumType -> JavaEnumType.wrap(enumType, typeMappings),
        mapType -> JavaMapType.wrap(mapType, typeMappings),
        JavaAnyType::javaAnyType);
  }
}
