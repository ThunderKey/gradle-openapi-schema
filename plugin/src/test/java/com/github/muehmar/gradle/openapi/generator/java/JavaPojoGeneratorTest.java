package com.github.muehmar.gradle.openapi.generator.java;

import static org.junit.jupiter.api.Assertions.assertEquals;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.Resources;
import com.github.muehmar.gradle.openapi.generator.data.Pojo;
import com.github.muehmar.gradle.openapi.generator.data.PojoMember;
import com.github.muehmar.gradle.openapi.generator.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.constraints.DecimalMax;
import com.github.muehmar.gradle.openapi.generator.constraints.DecimalMin;
import com.github.muehmar.gradle.openapi.generator.constraints.Max;
import com.github.muehmar.gradle.openapi.generator.constraints.Min;
import com.github.muehmar.gradle.openapi.generator.constraints.Pattern;
import com.github.muehmar.gradle.openapi.generator.constraints.Size;
import com.github.muehmar.gradle.openapi.generator.java.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.java.type.JavaTypes;
import com.github.muehmar.gradle.openapi.generator.settings.JsonSupport;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.writer.TestStringWriter;
import org.junit.jupiter.api.Test;

class JavaPojoGeneratorTest {

  @Test
  void generatePojo_when_minimalPojoSetting_then_correctPojoGenerated() {
    final TestStringWriter writer = new TestStringWriter();
    final JavaPojoGenerator pojoGenerator = new JavaPojoGenerator(() -> writer);

    final PojoSettings pojoSettings =
        new PojoSettings(
            JsonSupport.NONE,
            "com.github.muehmar",
            "Dto",
            false,
            false,
            PList.empty(),
            PList.empty());

    final Pojo pojo =
        new Pojo(
            "User",
            "User of the Application. This description is intentionally longer to see if its wrapped to a new line.",
            "Dto",
            PList.of(
                new PojoMember("id", "ID of this user", JavaType.ofName("long"), false),
                new PojoMember("name", "Name of this user", JavaTypes.STRING, false),
                new PojoMember(
                    "language",
                    "Preferred language of this user",
                    JavaType.javaEnum(PList.of("GERMAN", "ENGLISH")),
                    true)),
            false);

    pojoGenerator.generatePojo(pojo, pojoSettings);

    assertEquals(Resources.readString("/java/pojos/UserDtoMinimal.jv"), writer.asString().trim());
  }

  @Test
  void generatePojo_when_jsonSupportJackson_then_correctPojoGenerated() {
    final TestStringWriter writer = new TestStringWriter();
    final JavaPojoGenerator pojoGenerator = new JavaPojoGenerator(() -> writer);

    final PojoSettings pojoSettings =
        new PojoSettings(
            JsonSupport.JACKSON,
            "com.github.muehmar",
            "Dto",
            false,
            false,
            PList.empty(),
            PList.empty());

    final Pojo pojo =
        new Pojo(
            "User",
            "User of the Application",
            "Dto",
            PList.of(
                new PojoMember("id", "ID of this user", JavaType.ofName("long"), false),
                new PojoMember("name", "Name of this user", JavaTypes.STRING, false),
                new PojoMember(
                    "language",
                    "Preferred language of this user",
                    JavaType.javaEnum(PList.of("GERMAN", "ENGLISH")),
                    true)),
            false);

    pojoGenerator.generatePojo(pojo, pojoSettings);

    assertEquals(
        Resources.readString("/java/pojos/UserDtoJsonSupportJackson.jv"), writer.asString().trim());
  }

  @Test
  void generatePojo_when_enabledSafeBuilder_then_correctPojoGenerated() {
    final TestStringWriter writer = new TestStringWriter();
    final JavaPojoGenerator pojoGenerator = new JavaPojoGenerator(() -> writer);

    final PojoSettings pojoSettings =
        new PojoSettings(
            JsonSupport.NONE,
            "com.github.muehmar",
            "Dto",
            true,
            false,
            PList.empty(),
            PList.empty());

    final Pojo pojo =
        new Pojo(
            "User",
            "User of the Application",
            "Dto",
            PList.of(
                new PojoMember("id", "ID of this user", JavaType.ofName("long"), false),
                new PojoMember("name", "Name of this user", JavaTypes.STRING, false),
                new PojoMember(
                    "language",
                    "Preferred language of this user",
                    JavaType.javaEnum(PList.of("GERMAN", "ENGLISH")),
                    true)),
            false);

    pojoGenerator.generatePojo(pojo, pojoSettings);

    assertEquals(
        Resources.readString("/java/pojos/UserDtoEnabledSafeBuilder.jv"), writer.asString().trim());
  }

  @Test
  void generatePojo_when_enableConstraints_then_correctPojoGenerated() {
    final TestStringWriter writer = new TestStringWriter();
    final JavaPojoGenerator pojoGenerator = new JavaPojoGenerator(() -> writer);

    final PojoSettings pojoSettings =
        new PojoSettings(
            JsonSupport.NONE,
            "com.github.muehmar",
            "Dto",
            false,
            true,
            PList.empty(),
            PList.empty());

    final Pojo pojo =
        new Pojo(
            "User",
            "User of the Application",
            "Dto",
            PList.of(
                new PojoMember(
                    "id",
                    "ID of this user",
                    JavaType.ofName("long").withConstraints(Constraints.ofMax(new Max(50))),
                    false),
                new PojoMember(
                    "name",
                    "Name of this user",
                    JavaTypes.STRING.withConstraints(Constraints.ofSize(Size.of(10, 15))),
                    false),
                new PojoMember(
                    "lastName",
                    "Lastname of this user",
                    JavaTypes.STRING.withConstraints(Constraints.ofSize(Size.ofMin(10))),
                    false),
                new PojoMember(
                    "nickName",
                    "Nickname of this user",
                    JavaTypes.STRING.withConstraints(Constraints.ofSize(Size.ofMax(50))),
                    false),
                new PojoMember(
                    "email",
                    "Email of this user",
                    JavaTypes.STRING.withConstraints(Constraints.ofEmail()),
                    false),
                new PojoMember(
                    "height",
                    "Height of this user",
                    JavaTypes.DOUBLE.withConstraints(
                        Constraints.ofDecimalMin(new DecimalMin("120.0", true))
                            .withDecimalMax(new DecimalMax("199", false))),
                    false),
                new PojoMember(
                    "level",
                    "Level of this user",
                    JavaType.ofName("Long").withConstraints(Constraints.ofMin(new Min(5))),
                    true),
                new PojoMember(
                    "uppercase",
                    "Something uppercase",
                    JavaTypes.STRING.withConstraints(Constraints.ofPattern(new Pattern("[A-Z]"))),
                    true),
                new PojoMember(
                    "anotherPojo", "Another Pojo", JavaType.ofUserDefined("AnotherPojo"), true)),
            false);

    pojoGenerator.generatePojo(pojo, pojoSettings);

    assertEquals(
        Resources.readString("/java/pojos/UserDtoConstraints.jv"), writer.asString().trim());
  }
}
