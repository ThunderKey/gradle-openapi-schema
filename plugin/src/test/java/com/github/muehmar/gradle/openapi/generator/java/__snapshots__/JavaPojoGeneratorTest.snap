com.github.muehmar.gradle.openapi.generator.java.JavaPojoGeneratorTest.generatePojo_when_arrayPojoWithUniqueItems_then_correctPojoGenerated=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import javax.validation.constraints.AssertTrue;
import javax.validation.constraints.NotNull;

/**
 * Doses to be taken
 */
public class PosologyDto {
  @JsonValue
  private final List<Double> value;

  @JsonCreator
  public PosologyDto(
      List<Double> value
    ) {
    this.value = value;
  }

  /**
   * Doses to be taken
   */
  @NotNull
  public List<Double> getValue() {
    return value;
  }

  /**
   * Doses to be taken
   */
  public PosologyDto withValue(List<Double> value) {
    return new PosologyDto(value);
  }

  @AssertTrue(message = "value does not contain unique items")
  private boolean hasValueUniqueItems() {
    return new HashSet<>(value).size() == value.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final PosologyDto other = (PosologyDto) obj;
    return Objects.equals(this.value, other.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      value
    );
  }

  @Override
  public String toString() {
    return "PosologyDto{" +
      "value=" + value +
      "}";
  }
}
]